### DISCUSSION TITLE
Status

### DISCUSSION THREAD

**Discussion by kddnewton** [ruby/prism#29] [General]
This is a very general roadmap/status update that contains the latest information of the development of the new parser. The bullets are the nodes in syntax tree that loosely correspond to the nodes we're adding to this tree.

## Tokens

- [ ] Backref - We're not matching the semantics of when these should be created or not because we need to check `EXPR_FNAME`.
- [ ] Backtick - We're not matching the semantics of when this is handled because CRuby will check lex state and potentially create an `XStringBeg` node.
- [x] CHAR - `YP_TOKEN_CHARACTER_LITERAL`/`StringNode`
- [x] CVar - `YP_TOKEN_CLASS_VARIABLE`/`ClassVariableRead`/`ClassVariableWrite`
- [x] Comma - `YP_TOKEN_COMMA`
- [x] Comment - `yp_comment_t`
- [x] Const - `YP_TOKEN_CONSTANT`
- [x] EmbDoc - `yp_comment_t`
- [x] EmbVar - `YP_TOKEN_EMBVAR`
- [x] EndContent - `yp_comment_t`
- [x] FloatLiteral - `YP_TOKEN_FLOAT`/`FloatLiteral`
- [x] GVar - `YP_TOKEN_GLOBAL_VARIABLE`/`GlobalVariableRead`/`GlobalVariableWrite`
- [x] IVar - `YP_TOKEN_INSTANCE_VARIABLE`/`InstanceVariableRead`/`InstanceVariableWrite`
- [x] Imaginary - `YP_TOKEN_IMAGINARY_NUMBER`/`ImaginaryLiteral`
- [x] Int - `YP_TOKEN_INTEGER`/`IntegerLiteral`
- [x] Ident - `YP_TOKEN_IDENTIFIER`
- [x] Kw - This has been split out into individual keyword tokens for each keyword, which means: `YP_TOKEN_KEYWORD___ENCODING__`, `YP_TOKEN_KEYWORD___LINE__`, `YP_TOKEN_KEYWORD___FILE__`, `YP_TOKEN_KEYWORD_ALIAS`, `YP_TOKEN_KEYWORD_AND`, `YP_TOKEN_KEYWORD_BEGIN`, `YP_TOKEN_KEYWORD_BEGIN_UPCASE`, `YP_TOKEN_KEYWORD_BREAK`, `YP_TOKEN_KEYWORD_CASE`, `YP_TOKEN_KEYWORD_CLASS`, `YP_TOKEN_KEYWORD_DEF`, `YP_TOKEN_KEYWORD_DEFINED`, `YP_TOKEN_KEYWORD_DO`, `YP_TOKEN_KEYWORD_ELSE`, `YP_TOKEN_KEYWORD_ELSIF`, `YP_TOKEN_KEYWORD_END`, `YP_TOKEN_KEYWORD_END_UPCASE`, `YP_TOKEN_KEYWORD_ENSURE`, `YP_TOKEN_KEYWORD_FALSE`, `YP_TOKEN_KEYWORD_FOR`, `YP_TOKEN_KEYWORD_IF`, `YP_TOKEN_KEYWORD_IN`, `YP_TOKEN_KEYWORD_MODULE`, `YP_TOKEN_KEYWORD_NEXT`, `YP_TOKEN_KEYWORD_NIL`, `YP_TOKEN_KEYWORD_NOT`, `YP_TOKEN_KEYWORD_OR`, `YP_TOKEN_KEYWORD_REDO`, `YP_TOKEN_KEYWORD_RESCUE`, `YP_TOKEN_KEYWORD_RETRY`, `YP_TOKEN_KEYWORD_RETURN`, `YP_TOKEN_KEYWORD_SELF`, `YP_TOKEN_KEYWORD_SUPER`, `YP_TOKEN_KEYWORD_THEN`, `YP_TOKEN_KEYWORD_TRUE`, `YP_TOKEN_KEYWORD_UNDEF`, `YP_TOKEN_KEYWORD_UNLESS`, `YP_TOKEN_KEYWORD_UNTIL`, `YP_TOKEN_KEYWORD_WHEN`, `YP_TOKEN_KEYWORD_WHILE`, `YP_TOKEN_KEYWORD_YIELD`
- [x] LBrace - `YP_TOKEN_BRACE_LEFT`
- [x] LBracket - `YP_TOKEN_BRACKET_LEFT`
- [x] LParen - `YP_TOKEN_PARENTHESIS_LEFT`
- [x] Label - `YP_TOKEN_LABEL`
- [x] LabelEnd - `YP_TOKEN_LABEL_END`
- [x] Op - This has been split out into each individual operator, which means: `YP_TOKEN_AMPERSAND`, `YP_TOKEN_AMPERSAND_AMPERSAND`, `YP_TOKEN_AMPERSAND_AMPERSAND_EQUAL`, `YP_TOKEN_AMPERSAND_EQUAL`, `YP_TOKEN_BANG`, `YP_TOKEN_BANG_AT`, `YP_TOKEN_BANG_EQUAL`, `YP_TOKEN_BANG_TILDE`, `YP_TOKEN_BRACKET_LEFT_RIGHT`, `YP_TOKEN_CARET`, `YP_TOKEN_CARET_EQUAL`, `YP_TOKEN_COLON`, `YP_TOKEN_COLON_COLON`, `YP_TOKEN_DOT`, `YP_TOKEN_DOT_DOT`, `YP_TOKEN_DOT_DOT_DOT`, `YP_TOKEN_EQUAL`, `YP_TOKEN_EQUAL_EQUAL`, `YP_TOKEN_EQUAL_EQUAL_EQUAL`, `YP_TOKEN_EQUAL_GREATER`, `YP_TOKEN_EQUAL_TILDE`, `YP_TOKEN_GREATER`, `YP_TOKEN_GREATER_EQUAL`, `YP_TOKEN_GREATER_GREATER`, `YP_TOKEN_GREATER_GREATER_EQUAL`, `YP_TOKEN_KEYWORD_AND`, `YP_TOKEN_KEYWORD_NOT`, `YP_TOKEN_KEYWORD_OR`, `YP_TOKEN_LESS`, `YP_TOKEN_LESS_EQUAL`, `YP_TOKEN_LESS_EQUAL_GREATER`, `YP_TOKEN_LESS_LESS`, `YP_TOKEN_LESS_LESS_EQUAL`, `YP_TOKEN_MINUS`, `YP_TOKEN_MINUS_AT`, `YP_TOKEN_MINUS_EQUAL`, `YP_TOKEN_MINUS_GREATER`, `YP_TOKEN_PERCENT`, `YP_TOKEN_PERCENT_EQUAL`, `YP_TOKEN_PIPE`, `YP_TOKEN_PIPE_EQUAL`, `YP_TOKEN_PIPE_PIPE`, `YP_TOKEN_PIPE_PIPE_EQUAL`, `YP_TOKEN_PLUS`, `YP_TOKEN_PLUS_AT`, `YP_TOKEN_PLUS_EQUAL`, `YP_TOKEN_QUESTION_MARK`, `YP_TOKEN_SLASH`, `YP_TOKEN_SLASH_EQUAL`, `YP_TOKEN_STAR`, `YP_TOKEN_STAR_EQUAL`, `YP_TOKEN_STAR_STAR`, `YP_TOKEN_STAR_STAR_EQUAL`, `YP_TOKEN_TILDE`, `YP_TOKEN_TILDE_AT`
- [x] Period - `YP_TOKEN_DOT`
- [x] QSymbolsBeg - `YP_TOKEN_PERCENT_LOWER_I`
- [x] QWordsBeg - `YP_TOKEN_PERCENT_LOWER_W`
- [x] RBrace - `YP_TOKEN_BRACE_RIGHT`
- [x] RBracket - `YP_TOKEN_BRACKET_RIGHT`
- [x] RParen - `YP_TOKEN_PARENTHESIS_RIGHT`
- [x] RationalLiteral - `YP_TOKEN_RATIONAL_NUMBER`/`RationalLiteral`
- [x] RegexpBeg - `YP_TOKEN_REGEXP_BEGIN`
- [x] RegexpEnd - `YP_TOKEN_REGEXP_END`
- [ ] SymBeg - `YP_TOKEN_SYMBOL_BEGIN` - This is definitely not being handled correctly as it needs to differentiate itself from a ternary operator through a bunch of state that we don't currently track.
- [x] SymbolsBeg - `YP_TOKEN_PERCENT_UPPER_I`
- [ ] TLamBeg - `YP_TOKEN_LAMBDA_BEGIN` - We're not actually handling this correctly yet though, only in the cases where there are no parameters.
- [x] TLambda - `YP_TOKEN_MINUS_GREATER`
- [x] TStringBeg - `YP_TOKEN_STRING_BEGIN`
- [x] TStringContent - `YP_TOKEN_STRING_CONTENT`
- [x] TStringEnd - `YP_TOKEN_STRING_END`
- [x] WordsBeg - `YP_TOKEN_PERCENT_UPPER_W`

## Lists

- [x] ArrayLiteral - `ArrayNode`
- [x] QSymbols - `ArrayNode`
- [x] QWords - `ArrayNode`
- [x] Symbols - `ArrayNode`
- [x] Word - `StringNode`/`InterpolatedStringNode`
- [x] Words - `ArrayNode`

## Blocks

- [x] BEGINBlock - `PreExecutionNode`
- [x] BlockArg - `BlockParameterNode`
- [x] ClassDeclaration - `ClassNode`
- [x] Def - `DefNode`
- [x] DefEndless
- [x] Defs
- [x] ENDBlock - `PostExecutionNode`
- [x] KwRestParam - `KeywordRestParameterNode`
- [x] ModuleDeclaration - `ModuleNode`
- [x] Params - `ParametersNode`
- [x] RestParam - `RestParameterNode`
- [x] SClass - `SClassNode`

## Control flow

- [x] Begin
- [ ] BodyStmt
- [x] Break - `BreakNode`
- [ ] Case
- [x] Else - `ElseNode`
- [x] Elsif - `IfNode`
- [x] Ensure - `EnsureNode`
- [x] For - `ForNode`
- [x] If/IfMod - `IfNode`
- [x] IfOp - `Ternary`
- [x] Next - `NextNode`
- [x] Redo - `RedoNode`
- [x] Rescue/RescueEx/RescueMod - `RescueModifierNode`/`RescueNode`
- [x] Retry - `RetryNode`
- [x] Return/Return0 - `ReturnNode`
- [x] Super/ZSuper - `SuperNode`/`ForwardingSuperNode`
- [x] Unless/UnlessMod - `UnlessNode`
- [x] Until/UntilMod - `UntilNode`
- [ ] When
- [x] While/WhileMod - `WhileNode`
- [x] Yield/Yield0 - `YieldNode`

## Expressions

- [x] Alias - `AliasNode`
- [x] Assign - `CallNode`, `ClassVariableWrite`, `ConstantPathWriteNode`, `GlobalVariableWrite`, `InstanceVariableWrite`, `LocalVariableWrite`
- [x] Assoc - `AssocNode`
- [x] AssocSplat - `AssocSplatNode`
- [x] Binary - `CallNode`, `AndNode`, and `OrNode`
- [x] ConstPathField/ConstPathRef - `ConstantPathNode` and `ConstantPathWriteNode`
- [x] ConstRef - `ConstantRead`
- [x] Defined - `DefinedNode`
- [x] Dot2/Dot3 - `RangeNode`
- [x] Field - `CallNode`
- [x] HashLiteral - `HashNode`
- [ ] Lambda
- [ ] LambdaVar
- [ ] MAssign
- [x] MLHS - `MultiTargetNode`
- [ ] MLHSParen
- [ ] MRHS
- [x] Not - `CallNode`
- [x] OpAssign - `OperatorAssignmentNode`, `OperatorAndAssignmentNode`, and `OperatorOrAssignmentNode`
- [x] Paren - `ParenthesesNode`
- [x] Program - `Program`
- [x] Statements - `Statements`
- [ ] TopConstField/TopConstRef
- [x] VarAlias - `AliasNode`
- [x] VarField - `LocalVariableWrite`
- [x] VarRef - `TrueNode`, `FalseNode`, `NilNode`, and `LocalVariableRead`.
- [x] VoidStmt - Skipped
- [x] Unary - `UnaryNode`
- [x] Undef - `UndefNode`

## Strings

- [x] DynaSymbol - `SymbolNode` and `InterpolatedSymbolNode`
- [x] EmbExprBeg/EmbExprEnd - These are `StringInterpolated` nodes now.
- [ ] Heredoc
- [ ] HeredocBeg
- [ ] HeredocEnd
- [x] RegexpContent/RegexpLiteral - `RegularExpressionNode`/`InterpolatedRegularExpressionNode`
- [ ] StringConcat
- [x] StringContent - This is effectively now wrapped up into `StringNode`.
- [x] StringDVar - `GlobalVariableRead`/`ClassVariableRead`/`InstanceVariableRead`
- [x] StringEmbExpr - `StringInterpolatedNode`
- [x] StringLiteral - `StringNode` and `InterpolatedStringNode`
- [x] SymbolContent/SymbolLiteral - `SymbolNode`
- [x] XString/XStringLiteral - `XStringNode`/`InterpolatedXStringNode`

## Method calls

- [x] ARef/ARefField
- [x] ArgBlock
- [x] ArgParen
- [x] ArgStar
- [ ] Args
- [ ] ArgsForward
- [ ] BareAssocHash
- [ ] BlockVar
- [ ] BraceBlock
- [x] Call - `CallNode`
- [ ] Command
- [ ] CommandCall
- [ ] DoBlock
- [ ] ExcessedComma
- [x] FCall - `CallNode`
- [ ] MethodAddBlock
- [x] VCall - `CallNode`

## Pattern matching

- [ ] AryPtn
- [ ] FndPtn
- [ ] HshPtn
- [ ] In
- [ ] PinnedBegin
- [ ] PinnedVarRef
- [ ] RAssign

**Comment by iliabylich** [comment]
A few questions:

>  Backref - We're not matching the semantics of when these should be created or not because we need to check EXPR_FNAME.

Are you going to emulate those states in YARP? I'd really like to turn lexer into (almost) stateless module if possible.

>  Comment - We're creating the tokens but doing nothing to ignore them in the parser yet. We'll need to handle these by creating a separate list that we can optionally attach to the AST/serialized list later.

WDYT about emitting tokens separately? They do not participate in AST construction at all. In `lib-ruby-parser` I moved them to a separate field on parser result.

>  LabelEnd - We haven't handled this yet.

What is this token responsible for?

> Params

Please correct if I'm wrong but arguments is a part of the block/method declaration and parameters is what is passed to a method call, that's what I think I've in all parsers, but of course I can't find it explicitly stated anywhere. Is this node about method/block definition?

> DefEndless

I'd recommend to merge it with Def/Defs

> IfOp

What is this node?

> Rescue/RescueEx/RescueMod

Can it be a single `rescue` node?

>  Yield/Yield0

I understand the difference between `super`/`zsuper`, but are these nodes different?

> Assign

What is this node? Is for a setter call?

>  ConstPathField/ConstPathRef

> LambdaVar

What are these nodes?

> MLHS/MLHSParen

Can we merge these nodes into a single MLHS with optional begin/end locations?

> Paren

Is it a node type? Is it for empty parentheses? Can it be a Statements node?

> VarAlias

Is it for gvar/nthref alias? If so, can be it merged with Alias node?

> VoidStmt

Or is this node for empty parentheses? If so same question as above: Can it be a Statements node?

> EmbExprBeg

parser gem emits is as Begin (Statements in our case), can we do the same thing? Or is it a token type?

> ArgParen

> BlockVar

What is this node? Or is a token too?

> MethodAddBlock

Are we going to introduce these nodes at all?

Sorry if the list wasn't ready. Feel free to ignore this comment if so, I'll update it later.

**Reply by kddnewton** [reply]
You can find the definitions of most of the nodes here: https://ruby-syntax-tree.github.io/syntax_tree/. For the most part they mirror the names that come from ripper. If I've listed the nodes together on the same line, I want to merge them into individual nodes. I'm planning on collapsing a bunch more nodes as well.

Comments/EmbDoc should go into a separate list that can be optionally attached to the AST.

**Reply by kddnewton** [reply]
@iliabylich yeah I'd like it to be stateless as well, but unfortunately there's a couple of really complicated semantics we're going to have to mirror to get 100% compatibility. Check this out: https://github.com/tric/trick2022/blob/master/04-tompng/remarks.markdown.

**Reply by eregon** [reply]
> Please correct if I'm wrong but arguments is a part of the block/method declaration and parameters is what is passed to a method call, that's what I think I've in all parsers, but of course I can't find it explicitly stated anywhere. Is this node about method/block definition?

It's the opposite actually. A method/block definition has parameters. A call passes arguments to a method/block.
In Ruby docs the term arguments are often used for both, but when the term parameters is used it's always for the definition side, and notably the term was used about keyword-related changes in Ruby 3.0.
See https://en.wikipedia.org/wiki/Parameter_(computer_programming)#:~:text=Parameters%20appear%20in%20procedure%20definitions,an%20argument%20is%20an%20instance.

**Comment by kddnewton** [comment]
@eregon, @iliabylich, @enebo, @chrisseaton I've tagged v0.1.0 with a [status report](https://github.com/Shopify/yarp/releases/tag/v0.1.0) of everything I could think of. I've also updated this document with the current state of the project.

Obviously tons of stuff still to do, but I'm happy with the progress so far. If you have any glaring issues with the direction we're going in, let me know. Otherwise we'll keep trucking next week.

**Comment by kddnewton** [comment]
@eregon, @iliabylich, @enebo, @chrisseaton I've tagged v0.2.0 with a [status report](https://github.com/Shopify/yarp/releases/tag/v0.2.0) if you're interested.
